cpp/dolfinx/fem/interpolate.h

```diff
@@ -12,20 +12,188 @@
 #include "FiniteElement.h"
 #include "FunctionSpace.h"
 #include <dolfinx/common/IndexMap.h>
+#include <dolfinx/geometry/BoundingBoxTree.h>
+#include <dolfinx/geometry/utils.h>
 #include <dolfinx/mesh/Mesh.h>
 #include <functional>
 #include <numeric>
 #include <vector>
+#include <xtensor/xadapt.hpp>
 #include <xtensor/xarray.hpp>
 #include <xtensor/xtensor.hpp>
 #include <xtensor/xview.hpp>
 #include <xtl/xspan.hpp>
 
+namespace
+{
+
+/// @brief Send data from a set of processes to another set using neighborhod
+/// communicators.
+///
+/// The data to send is structured as (src_ranks.size(), value_size), where the
+/// ith row of `send_data` is sent to the process with rank `src_ranks[i]`. The
+/// ranks are using the ranks of the input communicator `comm`. `src_ranks` and
+/// `send_values` are assumed to be sorted, meaning that it is ordered by
+/// process to receive data. `dest_ranks` is a list of ranks the current process
+/// is receiving data from. This function returns a 2D array of shape
+/// (dest_ranks.size(), value_size). If the j-th dest rank is -1, then
+/// row(output,j) = (0,)*value_size.
+///
+/// @param[in] comm The mpi communicator
+/// @param[in] src_ranks The rank owning the values of each row in send_values
+/// @note It is assumed that src_ranks are already sorted.
+/// @param[in] dest_ranks The rank each local point is receiving data from.
+/// @note dest_ranks might contain -1 (no process owns the point)
+/// @param[in] send_values The values to send back to owner. Shape
+/// (src_ranks.size(), value_size).
+/// @returns An 2D array of shape (dest_ranks.size(), value_size) with values
+/// from the process owning the local point.
+template <typename T>
+xt::xtensor<T, 2> send_back_values(const MPI_Comm& comm,
```

Comment:
Function name isn't very informative and is too generic.

Is it intended to be in the public interface?

Body of the function is very similar to code elsewhere. Should it be made a general utility function?